### ISSUE TITLE
[Synapse Datasets + Collections] Define and incorporate schema for Synapse Datasets

### ISSUE THREAD

**Issue by aclayton555** [mc2-center/data-models#136]
Emerges from exploratory and feasibility analysis in: https://github.com/mc2-center/mc2-center-dcc/issues/71 

This ticket should track efforts to develop and implement a schema for annotating Synapse Datasets curated as part of the proposed MC2 Center workflow (Note that this is different from the existing 'Datasets' component in the MC2 Center data model). This is the first of several steps, which may be tracked in separate tickets as this work progress:

1) Define the schema (there have been ongoing discussions on this among the data managers group at Sage, but no resolution)
2) implementation as the JSON (this will be applicable and complimentary to ongoing efforts in NF)
3) exploration and incorporation of automation (again, pull in efforts from NF)
4) longer term: how this will look on the portal and what the expected user experience will be.

Prior work with Datasets in CSBC (precursor to MC2) and early MC2: https://sagebionetworks.jira.com/browse/CPO-313

**Comment by aclayton555** [comment]
24-9: @aditya-nath-sage will pick this up and chat with @jaybee84 about aligning approaches with NF.

Target output for this sprint: design doc for how to implement datasets across MC2 and NF (much of this captured in linked ticket above), including tentative annotation process. (will this leverage schematic or the Synapse API?)

Additional info on Synapse Datasets: https://help.synapse.org/docs/Datasets.2611281979.html

**Comment by aditya-nath-sage** [comment]
Goal is to create a design document for how MC2 and NF will want to handle this issue. Example design doc: https://docs.google.com/document/d/1dF1-FjGSdO3nkKArEsrnjnWFLeOV78MlvGZvM8smJVk/edit?pli=1#heading=h.47emx3tcx2wj

**Comment by aclayton555** [comment]
24-9 Close-Out: Currently working with DM group (at Sage) to create a org-wide Dataset schema. This may take some time to reach consensus, but we can prioritize incorporating a placeholder model that we can then add the finalized schema later. Check on this mid sprint in 24-10.

**Comment by aclayton555** [comment]
@aditya-nath-sage let's touch base on this during our check-in tomorrow!

**Comment by aclayton555** [comment]
Aditya and Orion to meet to align on this. In the meantime, @aditya-nath-sage to review ongoing design doc

Establish end of year goal for this effort

**Comment by aclayton555** [comment]
24-10: Orion has a rough script on how to bind entities in Synapse. Need to understand how the schematic outputs will work here and what the schema looks like.

**Comment by aclayton555** [comment]
24-11/12 Scope: Start working on this. This about how we surface datasets and collections that are on Synapse, and how these connect to publications via queryable metadata. Good to take stock of how many Datasets exist currently. Goal for end of sprint is a prelim design doc.

Another thought: for the record based datasets we have, how can we maybe generate and surface a collection of related datasets. SOme limitations here, as Synapse Collections currently only consolidate Dataset entities. One possibility is to generate entities from records, and create a Dataset from these, then create a Collection.

**Comment by Bankso** [comment]
Rough draft of a schema bind script: https://github.com/mc2-center/mc2-center-dcc/blob/add-utils-11-24/utils/synapse_json_schema_bind.py

Rough draft of a script to convert Synapse table info to annotations: https://github.com/mc2-center/mc2-center-dcc/blob/add-utils-11-24/utils/table_to_annotations.py

Script for creating a Synapse Dataset and adding entities from a folder: https://github.com/mc2-center/mc2-center-dcc/blob/add-utils-11-24/utils/build_datasets.py
- I imagine we could repurpose this to create Synapse Datasets using dataset link entities. To repurpose, we would want to add code (or write a separate script) where we 1) extract Dataset View table entries, likely on a project-by-project basis; 2) convert the Dataset link into a Synapse link entity (effectively a file) and store it in the 'datasets' folder 3) add the remaining Dataset View annotations to the link entity as Synapse annotations; 4) use 'build_datasets.py' to add link entities to Datasets. Step 4 could be modified to select for dataset link entities that have the same 'Study Key' or 'PublicationView Key', so only related dataset link entities are combined into a Synapse Dataset
- We should also consider how this relates to the 'DatasetView_id' attribute. Currently, we mint Synapse Ids to serve as entries in this field, by creating and immediately deleting folders. If we create link entities as part of our curation workflow, we could remove the minting step and just use the link Synapse Id instead. 
- Note that, when converting previously curated Dataset View entries to link entities, the entityId and DatasetView_id will end up being different, which may cause weirdness.

**Comment by aditya-nath-sage** [comment]
Updated the design doc slightly, but am working on understanding NF workflow completely and seeing how we can apply it in mc2. 

Will have a batter understanding of this by next sprint cycle

**Comment by aclayton555** [comment]
25-1 Close out - Aditya meeting with Christina to discuss this further this week. Approaching this as generalizable

**Comment by Bankso** [comment]
Some discussion in slack related to this:
https://sagebionetworks.slack.com/archives/C02Q4R41E4V/p1739522343520469
https://sagebionetworks.slack.com/archives/C02Q4R41E4V/p1739532754765319

- [Example of schema.org + croissant metadata for datasets](https://gist.github.com/susheel/e146a770f6edbfd10d915d0557332c31), provided by Susheel
- [Example of metadata conversion to croissant-style JSON](https://raw.githubusercontent.com/Sage-Bionetworks/core-models/refs/heads/main/data-models/sage-datasets/import-croissant.py), provided by Christina

More discussion will be needed to finalize the approach, but I think it is reasonable to conclude that:
1. we will be extracting metadata from Datasets, DatasetView, GrantView, Study, etc.
2. we will be binding schemas to Dataset entities
3. we will be applying annotations to Dataset entities
4. the schema and metadata will need to align with the Synapse Data Catalog standard, potentially this: https://www.synapse.org/Synapse:syn64765921/tables/
5. annotations should be automatically pulled into the Synapse Data Catalog, provided the appropriate `includedInDataCatalog` annotation is applied to the Dataset

Based on this, I think it makes sense to begin implementing this, as follows:
- Continue using Synapse Dataset entities
- Map existing metadata onto schema aligned with Synapse Data Catalog, which is based on schema.org Dataset schema
  - Incorporate `includedInDataCatalog` element, to reference the Synapse Data Catalog, but also the CCKP
  - Note: it would also be good to embed schema.org metadata into the CCKP, [example process is here](https://sagebionetworks.jira.com/browse/PORTALS-3342) and I assembled the [CCKP DataCatalog metadata here](https://github.com/mc2-center/data-models/blob/main/templates/cckp_datacatalog.jsonld)
- Bind Synapse Data Catalog JSON schema to Datasets
- Apply metadata to Datasets, depending on bound schema
  - note that attribute `includedInDataCatalog` should be set to the URL of the DataCatalog

Separately, consider the process described here: https://sagebionetworks.slack.com/archives/C02Q4R41E4V/p1739486550392299
Overview, assuming we apply the process above:
- Generate croissant metadata files (+/- schema.org elements) from existing metadata
- Store croissant metadata in a Synapse Folder
- Apply Synapse Data Catalog metadata from the source Datasets to croissant metadata files, meaning croissant files will be in the Synapse Data Catalog

Suggested croissant implementation (lower priority):
upload croissant metadata, but reference it in Dataset annotations, so the reference is pulled into the Synapse Data Catalog, but annotations need only be stored on a single entity

**Comment by Bankso** [comment]
Data catalog spreadsheet is available here: https://docs.google.com/spreadsheets/d/1zegC4jIGHu73JrUaSM6hwMrFpLbGymc4RgchncMYl7s/edit?gid=0#gid=0

columns include:
- community (MC2 Center)
- description
- contributors (institution)
- keywords
- individuals (number of samples/patients; available in Study metadata and Data Sharing Plans. Can also be determined based on Biospecimen/Model/Individual entries)
- link (link to entry in CCKP)
- croissant (link to croissant file in Synapse; we should be able to auto-generate these)

**Comment by aclayton555** [comment]
25-4: Working ongoing with Christina and DPE to apply and surface through data catalog. From MC2 perspective, need to decide how we want to interact in this ecosystem. Need to balance tables vs (Synapse) annotations so that information can be pulled into data catalog. As long as we know how to map our terms to those in the catalog, we should be okay. 

Continue to monitor this, but can visit in the context of MC2 toward end of Q2, aim to complete in Q3. Revisit for work in 25-6 sprint.

**Comment by aclayton555** [comment]
25-5: Still blocked? See what has been done for the NF data portal for croissant metadata on datasets. @aditya-nath-sage to talk with Christina and leverage script.

**Comment by aditya-nath-sage** [comment]
Christina worked with Brian Fauble to make the crossaint script more productionized.
Generating crossaint metadata from metadata manually updated in the NF portal happens through a script in this repo: https://github.com/Sage-Bionetworks-Workflows/orca-recipes
step 1: create datasets in synapse (make table with all dataset records so we can track and add crossaint links)
step 2: link portal dataset schema to each: [repo-prod.prod.sagebase.org//repo/v1/schema/type/registered/org.synapse.nf-portaldataset](http://repo-prod.prod.sagebase.org//repo/v1/schema/type/registered/org.synapse.nf-portaldataset)
step 3: manually fill out metadata necessary for crossaint processing
step 4: Run crossaint script on these datasets (talk to brian fauble for productionizing) to create crossaint files (https://github.com/Sage-Bionetworks-Workflows/orca-recipes/blob/main/dags/synapse-dataset-to-croissant.py)

Metadata needed: 
minimum: name, description, distribution (fileobjects, etc, this is updated automatically with brian's script right now)

**Comment by aclayton555** [comment]
25-5/6 mid sprint: Continue through 25-6 sprint (informs https://github.com/mc2-center/data-models/issues/208 ) and assess when to initiate related portal work in 25-7 sprint (tracked in https://sagebionetworks.jira.com/browse/CPO-1626).

Question(s): 
- do the croissant metadata need to live in a backend table? May be that this is generated as a separate file?
- for MC2, will this be limited to dataset entities?
- distribution - need to figure this out for datasets in Synapse vs repos like GEO
- as part of release, datsetView metadata are also applied as annotations, so these could also be leveraged?

Note: some dependencies here, keep @Bankso looped in

**Comment by aditya-nath-sage** [comment]
Update: need to update the annotationProperty schema (https://github.com/mc2-center/data-models/blob/main/modules/dataCatalog/annotationProperty.csv) to match the NF dataset collection schema (https://www.synapse.org/Synapse:syn50913342/datasets/). 

Once this is done, need to add the 48 or so datasets that are on Synapse to a dataset collection with the annotationProperty schema bound to it. Then we can use Bryan Fauble's script to create croissant metadata

**Comment by Bankso** [comment]
[Synapse Datasets listed on the CCKP are here](https://www.synapse.org/Synapse:syn21897968/tables/#query/eyJzcWwiOiJTRUxFQ1QgKiBGUk9NIHN5bjIxODk3OTY4IFdIRVJFIGRhdGFzZXRBbGlhcyBMSUtFICdzeW4lJyIsICJpbmNsdWRlRW50aXR5RXRhZyI6dHJ1ZSwgIm9mZnNldCI6MjUsICJsaW1pdCI6MjV9)

Not all have DatasetView metadata applied as annotations, but metadata is available in tables. Additional information not captured in DatasetView (individual count, etc.) will need to be captured separately. 

Note: these Datasets have been prepped while MC2 processes have been evolving, so there are some differences in file/record metadata and how Datasets for different projects have been put together. We have a standard approach now that we can apply to older Datasets, which will help us make sure things are consistent. 

Datasets that would be good to start with:
[CA261841 pathology](https://www.synapse.org/Synapse:syn21897968/tables/#query/eyJzcWwiOiJTRUxFQ1QgKiBGUk9NIHN5bjIxODk3OTY4IFdIRVJFIGRhdGFzZXRBbGlhcyBMSUtFICdzeW4lJyBBTkQgZ3JhbnROdW1iZXIgTElLRSAnJUNBMjYxODQxJSciLCAiaW5jbHVkZUVudGl0eUV0YWciOnRydWUsICJsaW1pdCI6MjV9): syn64713343, syn64713344, syn64779989, syn64780964
[CA263001 pathology](https://www.synapse.org/Synapse:syn21897968/tables/#query/eyJzcWwiOiJTRUxFQ1QgKiBGUk9NIHN5bjIxODk3OTY4IFdIRVJFIGRhdGFzZXRBbGlhcyBMSUtFICdzeW4lJyBBTkQgZ3JhbnROdW1iZXIgTElLRSAnJUNBMjYzMDAxJSciLCAiaW5jbHVkZUVudGl0eUV0YWciOnRydWUsICJsaW1pdCI6MjV9): syn63100169, syn63100173
[CA268072 microscopy and analysis](https://www.synapse.org/Synapse:syn21897968/tables/#query/eyJzcWwiOiJTRUxFQ1QgKiBGUk9NIHN5bjIxODk3OTY4IFdIRVJFIGRhdGFzZXRBbGlhcyBMSUtFICdzeW4lJyBBTkQgZ3JhbnROdW1iZXIgTElLRSAnJUNBMjY4MDcyJSciLCAiaW5jbHVkZUVudGl0eUV0YWciOnRydWUsICJsaW1pdCI6MjV9): syn61794747, syn61795436, syn61795441, syn61795455, syn61795458, syn61795461, syn61795464, syn61795539, syn61795662

**Comment by aclayton555** [comment]
25-8: Expecting to continue the curation aspects of this through 25-8. No anticipated blockers.

**Comment by Bankso** [comment]
I think we could use this [script from the croissant/data catalog work](https://github.com/Sage-Bionetworks/core-models/blob/main/data-models/sage-datasets/import-datasets.py) as a basis for how we generate data catalog entries

**Comment by Bankso** [comment]
@aditya-nath-sage I wanted to make sure we are aligned on approach for Dataset annotations and croissant metadata 

Following our Aug 28th meeting, here are my thoughts on next steps:
- align the [MC2 Data Catalog model](https://github.com/mc2-center/data-models/blob/main/modules/dataCatalog/annotationProperty.csv) with the [NF Portal Datasets model](https://github.com/nf-osi/nf-metadata-dictionary/blob/main/registered-json-schemas/PortalDataset.json)
  - these are similar already, since they are based off of schema.org/dataset, but the NF model has some additional fields and looks a bit more refined
  - consider how we want to align on enums/valid values - is there an ontology in use by NF that we can pull from?
- generate template and record Data Catalog entries for each Dataset in the [Dataset Collection](https://www.synapse.org/Synapse:syn68939725/datasets/) that @aditya-nath-sage generated
- using schematic, upload Data Catalog metadata to CCKP - MC2 Admin Synapse project + generate table
- generate a Synapse-compatible Data Catalog JSON schema and register it in Synapse
  - I believe this can be done using schematic
- bind the Data Catalog schema to each Dataset in the Dataset Collection using [synapse_json_schema_bind.py](https://github.com/mc2-center/mc2-center-dcc/blob/main/utils/synapse_json_schema_bind.py) (this is technically optional, since we will be pre-validating the metadata, but this has become a common step for shared Datasets)
- extract metadata from Data Catalog table in CCKP - MC2 Admin Synapse project and apply to Datasets as annotations using [table_to_annotations.py](https://github.com/mc2-center/mc2-center-dcc/blob/main/utils/table_to_annotations.py)
- apply the pipeline to build croissant metadata files to the Dataset Collection

Additional thoughts related to external dataset management:

- For GEO datasets, we could build the same kind of Data Catalog entries and perform the same annotation process using Synapse Link entities to represent the GEO records. This would require some secondary processing to get information about the files contained in the GEO dataset, which is required for croissant.
- One alternate idea for handling the GEO datasets is to create a Synapse Table that represents the dataset, then store the GEO/SRA file **information** there (one row = one file, pulled programmatically) and annotate the table with Data Catalog metadata (as above). The table could potentially be treated as a dataset by the croissant metadata generating pipeline.
- Storing file links, etc. in this table could potentially facilitate download through portals (related to [mc2-center-dcc/87](https://github.com/orgs/mc2-center/projects/4?pane=issue&itemId=104583152&issue=mc2-center%7Cmc2-center-dcc%7C87)). This could also provide a place where we associate additional metadata with GEO/SRA files, to ensure interoperability with data shared through Synapse, CRDC, etc.
- Last, if the table based dataset info management I described can be made compatible with the croissant pipeline, this could be a route for us to generate croissant metadata for other externally hosted files/datasets. This would be helpful for other projects at Sage that are interested in surfacing croissant metadata through portals (specifically, Bridge2AI)

**Comment by aditya-nath-sage** [comment]
As per Orion's comment above, we had a discussion with Christina Parry and Bryan Fauble about the Croissant work, have a plan moving forward and can create some croissant metadata from datasets we have on MC2 before site visit.

**Comment by aditya-nath-sage** [comment]
New DataCatalog model matching NF's model has been made,a nd bound to all datasets that are in [Croissant Pilot Dataset Collection](https://www.synapse.org/Synapse:syn68939725/datasets/). 
Annotations for all Datasets in collection have been added as well! Next step is to run Bryan Fauble's script and generate Croissant metadata for all datasets in the collection.

**Comment by aclayton555** [comment]
25-10 Kick Off: Closing this initial design and planning ticket for now. Continue efforts around the generation and deployment of the Croissant metadata in linked and new tickets.
